File: rust/src/env.rs
//! Environment for JSE execution with scope chaining.
//!
//! Following Python's env.py pattern:
//! - env has nullable parent field for scope chain lookup
//! - env provides register() method for def/defn
//! - env provides load() method for functor modules
//! - env.eval() delegates to ast.apply() (following jisp pattern)

use std::rc::Rc;
use std::cell::RefCell;
use std::collections::HashMap;
use serde_json::Value;
use crate::ast::{AstNode, AstError};

/// Functor type: function that takes env and arguments, returns Value
pub type Functor = fn(&Rc<RefCell<Env>>, &[Value]) -> Result<Value, AstError>;

/// Environment for JSE execution with scope chaining
pub struct Env {
    parent: Option<Rc<RefCell<Env>>>,
    bindings: HashMap<String, Value>,
    functor_map: HashMap<String, Functor>,
}

impl Env {
    /// Get this environment as an Rc<RefCell<Env>>.
    /// If this is the top-level environment, wrap it in a new Rc<RefCell>.
    /// Otherwise, return the parent Rc with this environment's bindings merged.
    /// For simplicity, we return the parent if available, or create a new wrapper.
    fn as_rc(&self) -> Rc<RefCell<Env>> {
        if let Some(parent) = &self.parent {
            Rc::clone(parent)
        } else {
            // For top-level env, we need to create a new Rc
            // This is a limitation - the caller should ideally pass the Rc in
            // But for our use case, we can work around it
            Rc::new(RefCell::new(self.clone()))
        }
    }
    /// Create a new empty environment
    pub fn new() -> Self {
        Self {
            parent: None,
            bindings: HashMap::new(),
            functor_map: HashMap::new(),
        }
    }

    /// Create a new environment with a parent (for closures/scope chain)
    pub fn new_with_parent(parent: Option<Rc<RefCell<Env>>>) -> Self {
        Self {
            parent,
            bindings: HashMap::new(),
            functor_map: HashMap::new(),
        }
    }

    /// Get the parent environment
    pub fn get_parent(&self) -> Option<Rc<RefCell<Env>>> {
        self.parent.as_ref().map(Rc::clone)
    }

    /// Resolve symbol to value by searching up the scope chain
    pub fn resolve(&self, symbol: &str) -> Option<Value> {
        if let Some(value) = self.bindings.get(symbol) {
            Some(value.clone())
        } else if let Some(parent) = &self.parent {
            parent.borrow().resolve(symbol)
        } else {
            None
        }
    }

    /// Register a new symbol binding (throws if exists)
    pub fn register(&mut self, name: String, value: Value) -> Result<(), AstError> {
        if self.bindings.contains_key(&name) {
            return Err(AstError::EvalError(
                format!("Symbol '{}' already exists in current scope", name)
            ));
        }
        self.bindings.insert(name, value);
        Ok(())
    }

    /// Set a symbol binding (overwrites if exists)
    pub fn set(&mut self, name: String, value: Value) {
        self.bindings.insert(name, value);
    }

    /// Check if symbol exists in scope chain
    pub fn exists(&self, name: &str) -> bool {
        self.bindings.contains_key(name) ||
        self.parent.as_ref().map_or(false, |p| p.borrow().exists(name))
    }

    /// Register a functor
    pub fn register_functor(&mut self, name: String, functor: Functor) {
        self.functor_map.insert(name, functor);
    }

    /// Load a module of functors into this environment
    pub fn load(&mut self, module: &HashMap<&'static str, Functor>) {
        for (name, functor) in module {
            self.functor_map.insert(name.to_string(), *functor);
        }
    }

    /// Evaluate an AST node in this environment.
    ///
    /// This is the core method that delegates to ast.apply(self),
    /// following the jisp pattern where env.eval() delegates to ast.apply().
    pub fn eval(&self, node: &dyn AstNode, env: &Rc<RefCell<Env>>) -> Result<Value, AstError> {
        node.apply(env)
    }

    /// Evaluate an AST node using self as the environment.
    /// This is a convenience method that creates an Rc wrapper if needed.
    pub fn eval_node(&self, node: &dyn AstNode) -> Result<Value, AstError> {
        let env = self.as_rc();
        node.apply(&env)
    }

    /// Apply a functor with AST nodes (for special forms)
    pub fn apply_functor(
        &self,
        name: &str,
        env: &Rc<RefCell<Env>>,
        args: &[&dyn AstNode]
    ) -> Result<Value, AstError> {
        let functor = self.resolve_functor(name)
            .ok_or_else(|| AstError::SymbolNotFound(name.to_string()))?;

        // Convert AST nodes to their JSON representation for special forms
        let json_args: Vec<Value> = args.iter()
            .map(|node| node.to_json())
            .collect();

        functor(env, &json_args)
    }

    /// Apply a functor with AST nodes using self as the environment.
    /// This is a convenience method that avoids passing env twice.
    pub fn apply_functor_node(&self, name: &str, args: &[&dyn AstNode]) -> Result<Value, AstError> {
        let env = self.as_rc();
        let functor = self.resolve_functor(name)
            .ok_or_else(|| AstError::SymbolNotFound(name.to_string()))?;

        let json_args: Vec<Value> = args.iter()
            .map(|node| node.to_json())
            .collect();

        functor(&env, &json_args)
    }

    /// Apply a functor with evaluated values
    pub fn apply_functor_with_values(
        &self,
        name: &str,
        env: &Rc<RefCell<Env>>,
        args: &[Value]
    ) -> Result<Value, AstError> {
        let functor = self.resolve_functor(name)
            .ok_or_else(|| AstError::SymbolNotFound(name.to_string()))?;

        functor(env, args)
    }

    /// Apply a functor with evaluated values using self as the environment.
    /// This is a convenience method that avoids passing env twice.
    pub fn apply_functor_values(&self, name: &str, args: &[Value]) -> Result<Value, AstError> {
        let env = self.as_rc();
        let functor = self.resolve_functor(name)
            .ok_or_else(|| AstError::SymbolNotFound(name.to_string()))?;

        functor(&env, args)
    }

    /// Resolve a functor from the environment chain
    pub fn resolve_functor(&self, name: &str) -> Option<Functor> {
        if let Some(&functor) = self.functor_map.get(name) {
            Some(functor)
        } else if let Some(parent) = &self.parent {
            parent.borrow().resolve_functor(name)
        } else {
            None
        }
    }
}

impl Clone for Env {
    fn clone(&self) -> Self {
        Self {
            parent: self.parent.as_ref().map(Rc::clone),
            bindings: self.bindings.clone(),
            functor_map: self.functor_map.clone(),
        }
    }
}

impl Default for Env {
    fn default() -> Self {
        Self::new()
    }
}

File: typescript/src/env.ts
/**
 * Environment for JSE execution.
 *
 * Follows the design in docs/regular.md:
 * - env has nullable parent field for scope chain lookup
 * - env provides register() method for def/defn
 * - env provides load() method for functor modules
 * - env.eval() delegates to ast.apply() (following jisp pattern)
 */

import type { JseValue, AstNodeLike } from "./types.js";

/**
 * Environment for JSE execution with scope chaining.
 */
export class Env {
  private _parent: Env | null;
  private _bindings: Map<string, JseValue>;

  constructor(parent: Env | null = null) {
    this._parent = parent;
    this._bindings = new Map();
  }

  /**
   * Get parent environment.
   */
  getParent(): Env | null {
    return this._parent;
  }

  /**
   * Resolve symbol to value by searching up the scope chain.
   * If symbol is not found in current environment, searches parent.
   * Returns undefined if symbol is not found anywhere in the chain.
   */
  resolve(symbol: string): JseValue | undefined {
    if (this._bindings.has(symbol)) {
      return this._bindings.get(symbol);
    }
    if (this._parent) {
      return this._parent.resolve(symbol);
    }
    return undefined;
  }

  /**
   * Register a new symbol binding in the current environment.
   * Throws if symbol already exists in current scope.
   * Used by $def and $defn operators.
   */
  register(name: string, value: JseValue): void {
    if (this._bindings.has(name)) {
      throw new Error(`Symbol '${name}' already exists in current scope`);
    }
    this._bindings.set(name, value);
  }

  /**
   * Set a symbol binding, overwriting if exists.
   * Unlike register(), this allows overwriting existing bindings.
   */
  set(name: string, value: JseValue): void {
    this._bindings.set(name, value);
  }

  /**
   * Check if symbol exists in current or parent scopes.
   */
  exists(name: string): boolean {
    if (this._bindings.has(name)) {
      return true;
    }
    if (this._parent) {
      return this._parent.exists(name);
    }
    return false;
  }

  /**
   * Evaluate an expression in this environment.
   * If expr is an AstNode, delegates to expr.apply(this).
   * Otherwise returns expr as-is (for primitive values).
   *
   * This follows the jisp pattern where env.eval() delegates to ast.apply().
   */
  eval(expr: JseValue): JseValue {
    // Check if it's an AstNode by checking for apply method
    if (
      typeof expr === "object" &&
      expr !== null &&
      "apply" in expr &&
      typeof (expr as AstNodeLike).apply === "function"
    ) {
      // This is an AstNode - delegate to its apply method
      return (expr as AstNodeLike).apply(this);
    }
    // Primitive value - return as-is
    return expr;
  }

  /**
   * Load functor module(s) into this environment.
   *
   * @param modules - One or more dictionaries of functor names and their implementations
   */
  load(...modules: Record<string, (env: Env, ...args: JseValue[]) => JseValue>[]): void {
    for (const module of modules) {
      for (const [name, functor] of Object.entries(module)) {
        this.register(name, functor);
      }
    }
  }
}

/**
 * Base environment interface for compatibility.
 * @deprecated Use Env class directly.
 */
export interface IEnv {
  resolve?(symbol: string): JseValue | undefined;
}

/**
 * Expression-only environment for basic and logic evaluation.
 * Can be extended to mount knowledge/statement data.
 */
export class ExpressionEnv extends Env implements IEnv {
  override resolve(_symbol: string): JseValue | undefined {
    return undefined;
  }
}
